### PR TITLE
chore: Add rust-toolchain file & adapt nix for changes

### PR DESCRIPTION
**Pull request by phated**
# Description

## Problem\*

Resolves 

## Summary\*

This adds a [rust-toolchain](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) file to the project.

I've also updated Nix flake to consume this file using the `Fenix` input instead of the `rust-overlay` input. This will bring the Nix setup closer to a "pure rustup" workflow.

I also took the chance to remove everything that existed for running tests since they don't actually work in Nix anymore.

And I changed the `default` package to be a combination of `nargo`, `noir_wasm`, and `noirc_abi_wasm`, which means running `nix build . -L` will produce a `result` directory that contains all of them. I think this should help with local testing of wasm since the tests often need artifacts from both `noir_wasm` and `noirc_abi_wasm`. I could also see us not making this the default but instead an `all` package—thoughts?

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: .github/workflows/test-integration.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ jobs:
         run: |
           nix build -L .#wasm
           mkdir -p ./.packages/noir_wasm
-          cp -r ./result/* ./.packages/noir_wasm/
+          cp -r ./result/noir_wasm/ ./.packages/noir_wasm/
```
Comment (phated) [inline .github/workflows/test-integration.yml:56]: @kobyhallx we should discuss these changes (the build-all-by-default nix change)

File: flake.nix [hunk 2]
```diff
@@ -284,143 +186,73 @@
...
 
           cargoArtifacts = native-cargo-artifacts;
         });
       };
 
       packages = {
-        default = noir-native;
+        # Nix flakes cannot build more than one derivation in one command (see https://github.com/NixOS/nix/issues/5591)
+        # so we use `symlinkJoin` to build everything as our default package.
+        default = pkgs.symlinkJoin { name = "noir"; paths = [ nargo noir_wasm noirc_abi_wasm ]; };
```
Comment (phated) [inline flake.nix:212]: cc @kobyhallx for thoughts on the build-all-by-default change

File: rust-toolchain.toml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+[toolchain]
+channel = "1.66.1"
+components = [ "rust-src" ]
+targets = [ "wasm32-unknown-unknown", "aarch64-apple-darwin" ]
```
Comment (kobyhallx) [inline rust-toolchain.toml:4]: Would we not need additional targets for linux?

File: flake.nix (hunk 2 above)
Comment (kobyhallx) [reply flake.nix]: let's make nargo as flake default so no extra args are needed to install into profile

File: .github/workflows/test-integration.yml (hunk 1 above)
Comment (kobyhallx) [reply .github/workflows/test-integration.yml]: If we have nargo as default, then we could utilize .#all to get all at the same time

File: rust-toolchain.toml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[toolchain]
```
Comment (kobyhallx) [inline rust-toolchain.toml:1]: Background https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file

File: rust-toolchain.toml (hunk 3 above)
Comment (phated) [reply rust-toolchain.toml:4]: The nice thing about the toolchain file is that is additive, so you'll always get your local system as a target and you'll also get these additional targets

**Comment by phated** [comment]
Wasm integration test run: https://github.com/noir-lang/noir/actions/runs/6226630424

**Review by kobyhallx — APPROVED** [review]
💯

**Pull request merged**